### ISSUE TITLE
Class for objective functional

### ISSUE THREAD

**Issue by moosmann** [odlgroup/odl#246]
Add module for objective functionals for primal-dual formulations of optimization problems.
Properties:
* convex conjugate
* proximal operator
* separable sum
* gradient

EDIT @kohr-h: gradient added

**Comment by adler-j** [comment]
What properties do we need?

**Comment by kohr-h** [comment]
As subclass of `Operator`, there are a couple of additional interesting properties, but basically none of those in the list above are needed in any situation.

For many cases, even evaluation of the functional itself is not required, but `Operator` forces one to write `_call`. So the abstract `ObjFunctional` class should implement `_call` as
```python
class ObjFunctional(Operator):
    ...
    def _call(self, x, out=None, **kwargs):
        raise NotImplementedError
```
just so `Operator` is happy. With some instructions for users added in the docs, in case they need the evaluation.

I could imagine some sort of `Solver` class as discussed in #76, which we may need anyway in the long run due to the desire to have an internal state. This solver would then tell in the doc which code paths need to be implemented for this guy to work, e.g. "I need obj_func.dual.prox and constr_func.prox." For users, some factory function which takes roughly the same input as now, would be needed. That function could create the required functionals internally and hand them to the class.

**Comment by adler-j** [comment]
Quick thought: Why do we need a class for this, how about simple duck-typing, perhaps adding these as "abstract" properties to `Operator`? Pro-cons for adding a new class?

**Comment by kohr-h** [comment]
I guess this depends a lot on the number of additional methods we're going to need, an I don't have a good guess yet.

An interesting issue is the way addition and scalar multiplication trickle through. That's why we've been talking about the special case of separable sums. The standard addition of functionals will yield the sum of the functionals with unknown `conjugate` and `proximal` since we cannot assume separability by default. The special `FunctionalSeparableSum` would take care of that (maybe implemented as a single-row `ProducSpaceOperator`?).

Scalar multiplication is also interesting since it propagates to `conjugate` and `proximal` in slightly non-obvious ways. @moosmann and @aringh worked out the expressions. That could either be added to the `Operator<Left, Right>ScalarMult` classes or in new guys.

We may also want a `bregman_dist` guy, possibly `subgradient`? Since all these methods/properties are special to functionals, I foresee a considerable cluttering of the basic `Operator` interface with methods which make no sense in most situations.

**Comment by adler-j** [comment]
Point of order: Isn't this issue equivalent to the ancient issue #39?

**Comment by kohr-h** [comment]
Absolutely correct. Some interesting thoughts in that issue.

**Comment by moosmann** [comment]
2nd thought about this issue: 
Regarding the Chambolle-Pock algorithm the purpose of introducing a functional class is that we want to evaluate the primal-dual gap in order to monitor the convergence of the algorithm which requires knowledge of the primal and dual objective functionals. However, these functionals typically contain indicator functions and instead of the 'full' prima-dual gap one only evaluates a 'partial' or 'conditional' primal-dual gap ignoring the indicator functions and monitors the conditions imposed by the indicator functions separately.

**Comment by kohr-h** [comment]
What exactly do you mean by "contain indicator functions"? Do they look like `f(x) + ind(x, C)` for some set `C`? What happens to conjugate and proximal in that case?

**Comment by aringh** [comment]
Yes, that is what is what @moosmann means. And the prox of an indicator function `ind(x, C)` is simply projection onto the set `C`, why I think #245 will be of interest here. One can also work out something relatively nice for convex conjugates, but I don't have it on top of my head right now (it should be some indicator function as well I suppose...).

The issue @moosmann refers to here is that if we have `f(x) = h(x) + ind(x, C)`, measuring a duality gap for `x not in C` is pointless since it will be `inf`. However, expressed in a slightly simplified way, one could measure the duality gap to `h(x)` instead. Then one cannot guarantee that it will be positive, but it will be a quantity that (monotonically?) goes to zero.

**Comment by kohr-h** [comment]
> And the prox of an indicator function ind(x, C) is simply projection onto the set C, why I think #245 will be of interest here.

I remembered something like that. What about the conjugate and proximal of `f(x) = h(x) + ind(x, C)`? Is that simply the sum of the individual contributions? If yes, we could just store the indicator functions separately, e.g. in an `indicator` property, and when calculating the conjugate/proximal, add those contributions.

**Comment by moosmann** [comment]
Some examples:
* if `G(x) = 0` with `x in X`, then `G^*(x)=ind(x, 0)` with `0` being the zero vector in `X`
* if `G(x) = ind(x, P)` where `P` is the set of non-negative vectors, then `G^*(x)=ind(-x, P)`
* if `f(x) = h(x) + ind(x, C)`, then `prox[f]` is the Euclidean projection onto `C`: `prox[f](x) = arg min_{y in C} ||x-y||_2^2`

**Comment by kohr-h** [comment]
> if f(x) = h(x) + ind(x, C), then prox[f] is the Euclidean projection onto C: prox[f](x) = arg min_{y in C} ||x-y||_2^2

Where did `h` go? That's the most interesting question here.

**Comment by moosmann** [comment]
Oops, copy-and-paste error. It should have been:
if `f(x) = ind(x, C)`, then `prox[f]` is the Euclidean projection onto `C`: `prox[f](x) = arg min_{y in C} ||x-y||_2^2`
I guess for `f(x) = h(x) + ind(x, C)`, all you can say is `prox[f](x) = arg min_{y in C} (h(x) + ||x-y||_2^2)`

**Comment by kohr-h** [comment]
That's what I thought, unfortunately. So we need one guy for each combination of `h` and `C` where the `conjugate` and `proximal` have to be calculated by hand. Regarding the primal-dual gap, we could instead add a property like `finite_part` where only the finite part of the objective is evaluated.

**Comment by moosmann** [comment]
Exactly. The question is what do we demand from the user to be implemented? If we only need the finite parts of the objectives in order to evaluate the primal-dual gap, I think we shouldn't force the user to implement the full objective and to think about how to implement an indicator function which will never be evaluated. What do you think?

**Comment by kohr-h** [comment]
If we really NEVER need the indicator in the functional itself, we could advise the user not to add that part in the implementation for the primal-dual gap to work properly. It will be needed in the conjugate and proximal operators, though, and here we should make clear that it's important to properly implement these in the presence of indicators.

Question: are we sure that the indicators are really never needed in the functional?

**Comment by aringh** [comment]
In what I can see in the articles if are looking at, the indicator function itself will never be used. The probability that other cost functions used in Chambolle-Pock will need them (so not only L2 or L1 norm) seems slim: the point of this algorithm is that the prox-operators should have explicit forms so that the updates are fast to compute.

However, the `ObjFunctional` class should be possible to use in basically all optimization solvers. And in other solvers there might be a point of having indicator functions (think of a Newton solver on a problem with, for example, with a positivity constraint. An indicator function can be useful in a step length computation, order to limit the step length of the Newton solver, and in that way always keep the components positive.).

Maybe the `finite_part` could be a nice trade-off between the two? Then one could have one duality gap which takes the actual functional and its dual (if that is what the user wants), and one 'conditional gap', which uses the `finite_part`.

**Comment by aringh** [comment]
> What exactly do you mean by "contain indicator functions"? Do they look like f(x) + ind(x, C) for some set C? What happens to conjugate and proximal in that case?

I don't have the complete answer (and I don't think there is a simple answer to the general question), but for reference:
* the prox of the indicator function of a set C is simply the projection onto the set C.
* the convex conjugate of a indicator function of a set C is, by definition, the [support function](https://en.wikipedia.org/wiki/Support_function) of the set C.

**Comment by aringh** [comment]
Closed by  #498
